### PR TITLE
external: parallel seekPropsOffsets

### LINKED ISSUE #51639: global sort 120T test seekPropsOffsets performance problems
## Enhancement

When `seekPropsOffsets` need to handle 2800 files

```
[reader.go:43] [arguments] [data-file-count=2800] [stat-file-count=2800] [start-key=dIAAAAAAAABoX3IBNzAxMzk2ODj/OTFzSlFORW3/SzE4ckM0Nk//RVY1WFZzYTH/dDFZcGYwQVT/a1kAAAAAAAD5] [end-key=dIAAAAAAAABoX3IBNzAxNDEwOTf/N1ZmbVhBVTf/TENBUjlxYlP/Q1Z0ZjBIOWj/SlRWWU5uVm3/WgAAAAAAAAD4]
```

The single thread heap-based merge sort becomes bottleneck

image
image

as @wjhuang2016 suggested, we should find a more parallel way to handle it, like readAllData

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #51639

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
   AWS 40T ingest step becomes faster
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/util.go [hunk 1]
```diff
@@ -42,68 +44,109 @@ func seekPropsOffsets(
...
-	multiFileStat := MultipleFilesStat{Filenames: make([][2]string, 0, len(paths))}
-	for _, path := range paths {
-		multiFileStat.Filenames = append(multiFileStat.Filenames, [2]string{"", path})
+	offsetsPerFile := make([][]uint64, len(paths))
+	for i := range offsetsPerFile {
+		offsetsPerFile[i] = make([]uint64, len(starts))
 	}
-	iter, err := NewMergePropIter(ctx, []MultipleFilesStat{multiFileStat}, exStorage, checkHotSpot)
-	if err != nil {
+	// if **all** files' first key is smaller than start key, we may lost the data
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/util.go:58]:
```suggestion
	// if **all** files' first keys are smaller than start key, we may lost the data
```

File: br/pkg/lightning/backend/external/util.go [hunk 2]
```diff
@@ -42,68 +44,109 @@ func seekPropsOffsets(
...
+			}
+			defer r.Close()
+
+			moved := false
+			keyIdx := 0
+			curKey := starts[keyIdx]
+
+			p, err3 := r.nextProp()
+			for {
+				if egCtx.Err() != nil {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/util.go:81]: /hold

contention is high

File: br/pkg/lightning/backend/external/util.go [hunk 3]
```diff
@@ -42,68 +44,105 @@ func seekPropsOffsets(
...
-	for _, path := range paths {
-		multiFileStat.Filenames = append(multiFileStat.Filenames, [2]string{"", path})
+	offsetsPerFile := make([][]uint64, len(paths))
+	for i := range offsetsPerFile {
+		offsetsPerFile[i] = make([]uint64, len(starts))
 	}
-	iter, err := NewMergePropIter(ctx, []MultipleFilesStat{multiFileStat}, exStorage, checkHotSpot)
-	if err != nil {
+	// if **all** files' first key is smaller than start key, we may lost the data
+	// between [start key, min(first key of all files)). So record this case for each
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/util.go:59]: larger? if all `first key in file` < `start key`, this range is invalid

File: br/pkg/lightning/backend/external/util.go [hunk 4]
```diff
@@ -42,68 +44,105 @@ func seekPropsOffsets(
...
+			curKey := starts[keyIdx]
+
+			p, err3 := r.nextProp()
+			for {
+				switch err3 {
+				case nil:
+				case io.EOF:
+					for keyIdx++; keyIdx < len(starts); keyIdx++ {
+						offsetsPerFile[i][keyIdx] = offsetsPerFile[i][keyIdx-1]
+					}
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/util.go:86]:
```suggestion
					currFileoffset := offsetsPerFile[i][keyIdx]
					for keyIdx++; keyIdx < len(starts); keyIdx++ {
						offsetsPerFile[i][keyIdx] = currFileoffset
					}
```
i think this more clear

File: br/pkg/lightning/backend/external/util.go (hunk 3 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/external/util.go:59]: why data lost, shouldn't we allow some file doesn't have kv in some range?

Comment (lance6716) [reply br/pkg/lightning/backend/external/util.go]: https://github.com/pingcap/tidb/pull/51676/commits/82a98c95f14984fdc9814e982ad038b5fb37526c I have added some comments, please check if it's clear enough.

We can allow part of files does not cover the first `starts` key, but not all files.

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**